src/mono/mono/mini/debugger-agent.c

```diff
@@ -5466,6 +5007,17 @@ buffer_add_value_full (Buffer *buf, MonoType *t, void *addr, MonoDomain *domain,
 
 		if (!obj) {
 			buffer_add_byte (buf, VALUE_TYPE_ID_NULL);
+			if (CHECK_PROTOCOL_VERSION (3, 0)) {
+				buffer_add_byte (buf, t->type);
+				if (t->type == MONO_TYPE_CLASS || t->type == MONO_TYPE_STRING)
+					buffer_add_typeid (buf, domain, mono_class_from_mono_type_internal (t));
+				if (t->type == MONO_TYPE_SZARRAY || t->type == MONO_TYPE_ARRAY) {
+					buffer_add_byte(buf, m_class_get_byval_arg (m_class_get_element_class (mono_class_from_mono_type_internal (t)))->type);
+					buffer_add_int (buf, m_class_get_rank (mono_class_from_mono_type_internal (t)));
+					if (m_class_get_byval_arg (m_class_get_element_class (mono_class_from_mono_type_internal (t)))->type == MONO_TYPE_CLASS)
+						buffer_add_typeid (buf, domain, m_class_get_element_class (mono_class_from_mono_type_internal (t)));
+				}
```

Comment:
this might be clearer with a `switch (t->type)`.
Will you need to handle other cases - generic instances, pointers types?  Do you need to check `t->byref` ?

Reply:
I think so, for icordbg I need info of the object even when it's a null value. I refactored it into a new function and using the switch as you suggested.